File: rust/tests/test_formula_escaping.rs
// TC2 — Formula-value escaping. Filter predicates build formulas by interpolating user values into
// quoted string literals; a value containing a quote, backslash, or other meta-character must be
// escaped or the formula breaks (or silently mis-matches). These tests round-trip special-character
// values through storage AND through the generated `.equals()/.contains()` filter DSL against the
// live base, proving the escaping is correct. Parity target for the other 8 suites.

use myairtable_tests::*;

fn setup() -> Airtable {
    dotenvy::from_filename(".env").ok();
    let api_key = std::env::var("AIRTABLE_API_KEY").expect("AIRTABLE_API_KEY must be set");
    let base_id = std::env::var("AIRTABLE_BASE_ID").expect("AIRTABLE_BASE_ID must be set");
    Airtable::new(&api_key, &base_id)
}

// A primary-key value unique to this run (distinct per scenario via `label`).
fn primary_key(suite: &str, label: &str) -> String {
    let ts = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis();
    let rand = ts.wrapping_mul(2654435761) % 900000 + 100000;
    format!("Rust {suite} {label} {ts}-{rand}")
}

// Values that break naive string interpolation into an Airtable formula literal.
// Newline is covered separately (storage only) — Airtable formula string literals can't carry a
// raw newline, so it's a storage concern, not a filter one.
fn special_values() -> Vec<&'static str> {
    vec![
        "O'Brien single quote",
        "say \"hi\" double quote",
        "back\\slash path",
        "trailing backslash\\",
        "mixed a\"b'c\\d",
        "unicode café ☕ 日本語 🎉",
    ]
}

// Best-effort cleanup: ignore failures (the live base tolerates leftover records).
async fn try_delete(at: &Airtable, id: &str) {
    let _ = at.primary.delete_one(&id.to_string()).await;
}

// =============================================================================
// Eq filter matches a value with special characters
// =============================================================================

#[tokio::test]
async fn eq_filter_matches_value_with_special_chars() {
    let at = setup();
    let f = PrimaryModel::F;

    for special in special_values() {
        let suite = primary_key("Escaping", "Eq");

        let created = at
            .primary
            .create_one(&PrimaryModel {
                primary_key: Some(suite.clone()),
                single_line_text: Some(special.to_string()),
                ..Default::default()
            })
            .await
            .unwrap();
        let record_id = created.id.as_deref().unwrap().to_string();

        // 1. Storage round-trips the exact value.
        assert_eq!(
            created.single_line_text.as_deref(),
            Some(special),
            "storage did not round-trip value: {special:?}"
        );

        // 2. The .equals() filter, scoped to this record, matches via correctly-escaped interpolation.
        let scope = format!("FIND(\"{suite}\", {{Primary Key}})");
        let filter = formula::AND(&[&scope, &f.single_line_text.equals(special, true, false)]);
        let params = AirtableQuery::new().formula(filter);
        let results = at.primary.get_many(&params).await.unwrap();

        assert!(
            results
                .iter()
                .any(|r| r.id.as_deref() == Some(record_id.as_str())),
            "Eq filter did not match record for value: {special:?}"
        );
        for r in &results {
            assert_eq!(
                r.single_line_text.as_deref(),
                Some(special),
                "Eq filter matched a record with wrong value for: {special:?}"
            );
        }

        try_delete(&at, &record_id).await;
    }
}

// =============================================================================
// Contains filter matches a value with special characters
// =============================================================================

#[tokio::test]
async fn contains_filter_matches_value_with_special_chars() {
    let at = setup();
    let f = PrimaryModel::F;

    for special in special_values() {
        let suite = primary_key("Escaping", "Contains");
        // Embed the special token inside a longer value so Contains (FIND>0) is a real substring test.
        let stored = format!("prefix {special} suffix");

        let created = at
            .primary
            .create_one(&PrimaryModel {
                primary_key: Some(suite.clone()),
                single_line_text: Some(stored.clone()),
                ..Default::default()
            })
            .await
            .unwrap();
        let record_id = created.id.as_deref().unwrap().to_string();

        let scope = format!("FIND(\"{suite}\", {{Primary Key}})");
        let filter = formula::AND(&[&scope, &f.single_line_text.contains(special, true, false)]);
        let params = AirtableQuery::new().formula(filter);
        let results = at.primary.get_many(&params).await.unwrap();

        assert!(
            results
                .iter()
                .any(|r| r.id.as_deref() == Some(record_id.as_str())),
            "Contains filter did not match record for value: {special:?}"
        );

        try_delete(&at, &record_id).await;
    }
}

// =============================================================================
// Special characters in the primary key itself, matched with .equals()
// =============================================================================

#[tokio::test]
async fn special_characters_in_primary_key_round_trip_and_filter() {
    let at = setup();
    let f = PrimaryModel::F;

    let suite = primary_key("Escaping", "PK");
    let pk = format!("{suite} O'Brien \"quote\" back\\slash");

    let created = at
        .primary
        .create_one(&PrimaryModel {
            primary_key: Some(pk.clone()),
            ..Default::default()
        })
        .await
        .unwrap();
    let record_id = created.id.as_deref().unwrap().to_string();

    assert_eq!(created.primary_key.as_deref(), Some(pk.as_str()));

    let params = AirtableQuery::new().formula(f.primary_key.equals(&pk, true, false));
    let results = at.primary.get_many(&params).await.unwrap();

    assert!(
        results
            .iter()
            .any(|r| r.id.as_deref() == Some(record_id.as_str())),
        "Eq filter did not match record with special chars in primary key"
    );
    assert_eq!(
        results.len(),
        1,
        "Eq on primary key should match exactly one record"
    );

    try_delete(&at, &record_id).await;
}

// =============================================================================
// Newline / tab values round-trip through storage (long text)
// =============================================================================

#[tokio::test]
async fn newline_and_tab_values_round_trip_through_storage() {
    let at = setup();

    // Newlines/tabs are a storage concern (long text), not expressible in a formula literal.
    // Verify they survive create -> fetch unchanged.
    let suite = primary_key("Escaping", "Newline");
    let value = "line1\nline2\twith tab\r\nwindows";

    let created = at
        .primary
        .create_one(&PrimaryModel {
            primary_key: Some(suite),
            long_text: Some(value.to_string()),
            ..Default::default()
        })
        .await
        .unwrap();
    let record_id = created.id.as_deref().unwrap().to_string();

    let fetched = at.primary.get_one(&record_id).await.unwrap();
    assert_eq!(
        fetched.long_text.as_deref(),
        Some(value),
        "newline/tab value did not round-trip through storage"
    );

    try_delete(&at, &record_id).await;
}

File: python/tests/test_formula_escaping.py
"""TC2 - Formula-value escaping.

Filter predicates build formulas by interpolating user values into quoted string
literals; a value containing a quote, backslash, or other meta-character must be
escaped or the formula breaks (or silently mis-matches). These tests round-trip
special-character values through storage AND through the generated ``.eq()``/
``.contains()`` filter DSL against the live base, proving the escaping is correct.
Parity target mirrors csharp/tests/TestFormulaEscaping.cs.
"""

import random
import time

import pytest
from pyairtable import formulas as F  # noqa: N812

from output import AND, Airtable, PrimaryModel

# Values that break naive string interpolation into an Airtable formula literal.
# Newline is covered separately (storage only) - Airtable formula string literals
# can't carry a raw newline, so it's a storage concern, not a filter one.
SPECIAL_VALUES = [
    "O'Brien single quote",
    'say "hi" double quote',
    "back\\slash path",
    "trailing backslash\\",
    "mixed a\"b'c\\d",
    "unicode café ☕ 日本語 🎉",
]


@pytest.fixture(scope="module")
def airtable() -> Airtable:
    return Airtable()


def _primary_key(suite: str, label: str) -> str:
    """A primary-key value unique to this run (distinct per scenario via label)."""
    ts = int(time.time() * 1000)
    rand = random.randint(100000, 999999)
    return f"Python {suite} {label} {ts}-{rand}"


def _scope(suite: str) -> F.Formula:
    """FIND("<suite>", {Primary Key}) - scopes a filter to records from this run."""
    return F.FIND(suite, F.Field("Primary Key"))


def _try_delete(airtable: Airtable, record_id: str | None) -> None:
    if not record_id:
        return
    try:
        airtable.primary.delete(record_id=record_id)
    except Exception:
        # best-effort cleanup
        pass


@pytest.mark.parametrize("special", SPECIAL_VALUES)
def test_eq_filter_matches_value_with_special_chars(airtable: Airtable, special: str) -> None:
    suite = _primary_key("Escaping", "Eq")
    model = PrimaryModel()
    model.primary_key = suite
    model.single_line_text = special
    created = airtable.primary.create(model)
    record_id = created.id
    try:
        # 1. Storage round-trips the exact value.
        assert created.single_line_text == special

        # 2. The .eq() filter, scoped to this record, matches via correctly-escaped
        #    interpolation.
        formula = AND(_scope(suite), PrimaryModel.f.single_line_text.eq(special))
        results = airtable.primary.get(formula=formula)
        assert any(r.id == record_id for r in results)
        for r in results:
            assert r.single_line_text == special
    finally:
        _try_delete(airtable, record_id)


@pytest.mark.parametrize("special", SPECIAL_VALUES)
def test_contains_filter_matches_value_with_special_chars(airtable: Airtable, special: str) -> None:
    suite = _primary_key("Escaping", "Contains")
    # Embed the special token inside a longer value so contains (FIND>0) is a real
    # substring test.
    stored = "prefix " + special + " suffix"
    model = PrimaryModel()
    model.primary_key = suite
    model.single_line_text = stored
    created = airtable.primary.create(model)
    record_id = created.id
    try:
        formula = AND(_scope(suite), PrimaryModel.f.single_line_text.contains(special))
        results = airtable.primary.get(formula=formula)
        assert any(r.id == record_id for r in results)
    finally:
        _try_delete(airtable, record_id)


def test_special_characters_in_primary_key_round_trip_and_filter(airtable: Airtable) -> None:
    # The primary key itself carries special chars, and we match on it with .eq().
    suite = _primary_key("Escaping", "PK")
    pk = suite + ' O\'Brien "quote" back\\slash'
    model = PrimaryModel()
    model.primary_key = pk
    created = airtable.primary.create(model)
    record_id = created.id
    try:
        assert created.primary_key == pk
        formula = PrimaryModel.f.primary_key.eq(pk)
        results = airtable.primary.get(formula=formula)
        assert any(r.id == record_id for r in results)
        assert len(results) == 1
    finally:
        _try_delete(airtable, record_id)


def test_newline_and_tab_values_round_trip_through_storage(airtable: Airtable) -> None:
    # Newlines/tabs are a storage concern (long text), not expressible in a formula
    # literal. Verify they survive create -> fetch unchanged.
    suite = _primary_key("Escaping", "Newline")
    value = "line1\nline2\twith tab\r\nwindows"
    model = PrimaryModel()
    model.primary_key = suite
    model.long_text = value
    created = airtable.primary.create(model)
    record_id = created.id
    try:
        fetched = airtable.primary.get(record_id)
        assert fetched.long_text == value
    finally:
        _try_delete(airtable, record_id)
